src/screens/PinAuthenticationScreen.js

```diff
@@ -0,0 +1,198 @@
+import { useRoute } from '@react-navigation/core';
+import React, { useCallback, useEffect, useRef, useState } from 'react';
+import { Alert } from 'react-native';
+import FastImage from 'react-native-fast-image';
+import styled from 'styled-components/primitives';
+import RainbowLogo from '../assets/rainbows/light.png';
+import { Centered, Column, ColumnWithMargins } from '../components/layout';
+import { Numpad, PinValue } from '../components/numpad';
+import { SheetTitle } from '../components/sheet';
+import {
+  getAuthTimelock,
+  saveAuthTimelock,
+} from '../handlers/localstorage/globalSettings';
+import { useDimensions, useShakeAnimation } from '../hooks';
+import { useBlockBackButton } from '../hooks/useBlockBackButton';
+import { useNavigation } from '../navigation/Navigation';
+import { colors, padding } from '@rainbow-me/styles';
+
+const Logo = styled(FastImage).attrs({
+  source: RainbowLogo,
+})`
+  width: 80;
+  height: 80;
+`;
+
+const MAX_ATTEMPTS_LEFT = 10;
+const TIMELOCK_INTERVAL_MINUTES = 5;
+
+const PinAuthenticationScreen = () => {
+  useBlockBackButton();
+  const { params } = useRoute();
+  const { goBack, setParams } = useNavigation();
+  const [errorAnimation, onShake] = useShakeAnimation();
+
+  const { isNarrowPhone, isSmallPhone, isTallPhone } = useDimensions();
+
+  const [attemptsLeft, setAttemptsLeft] = useState(MAX_ATTEMPTS_LEFT);
+  const [value, setValue] = useState('');
+  const [initialPin, setInitialPin] = useState('');
+  const [actionType, setActionType] = useState(
+    params.validPin ? 'authentication' : 'creation'
+  );
+
+  const finished = useRef(false);
+
+  useEffect(() => {
+    return () => {
+      if (!finished.current) {
+        params.onCancel();
+      }
+    };
+  }, [params, setParams]);
+
+  useEffect(() => {
+    const checkTimelock = async () => {
+      // When opening the screen we need to check
+      // if the user wasn't banned for too many tries
+      const timelock = await getAuthTimelock();
+      if (timelock) {
+        const stillBanned = Date.now() < timelock;
```

Comment:
Can reuse Date.now() (on line 60 and 62) and can set timeLeftMS to timelock - dateNow (instead of dateNow - timelock) so you don't need the Math.abs later